File: aura-impl/src/main/java/org/auraframework/impl/controller/LabelController.java
/*
 * Copyright (C) 2013 salesforce.com, inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.auraframework.impl.controller;

import static org.auraframework.instance.AuraValueProviderType.LABEL;

import javax.inject.Inject;

import org.auraframework.annotations.Annotations.ServiceComponent;
import org.auraframework.ds.servicecomponent.GlobalController;
import org.auraframework.expression.PropertyReference;
import org.auraframework.impl.expression.PropertyReferenceImpl;
import org.auraframework.instance.GlobalValueProvider;
import org.auraframework.service.ContextService;
import org.auraframework.system.Annotations.ActionGroup;
import org.auraframework.system.Annotations.AuraEnabled;
import org.auraframework.system.Annotations.Key;
import org.auraframework.throwable.quickfix.QuickFixException;

@ServiceComponent
public class LabelController implements GlobalController {
    private static final String NAME = "aura://LabelController";

    @Inject
    private ContextService contextService;

    @AuraEnabled
    @ActionGroup(value = "aura")
    public String getLabel(@Key("section") String section, @Key("name") String name) throws QuickFixException {
        GlobalValueProvider labelProvider = contextService.getCurrentContext().getGlobalProviders()
                .get(LABEL.getPrefix());
        PropertyReference labelRef = new PropertyReferenceImpl(section + "." + name, null);
        return (String) labelProvider.getValue(labelRef);
    }

    @Override
    public String getQualifiedName() {
        return NAME;
    }
}

File: aura-components/src/main/components/aura/label/labelController.js
/*
 * Copyright (C) 2013 salesforce.com, inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
({
    updateTemplate:function(cmp){
        if(!cmp.get("v.updating")){
            var bodyTemplate=cmp.get("v.body");
            for(var i=0;i<bodyTemplate.length;i++){
                if (bodyTemplate[i].isInstanceOf("aura:text")) {
                    bodyTemplate.splice(i--,1);
                }
            }
            cmp.set("v.template", bodyTemplate);
        }
        cmp.set("v.updating",false);
    },

    update:function(cmp){
        var value=cmp.get("v.value")||'';
        // No value, clear out body
        if(!value||!$A.util.isString(value)){
            cmp.set("v.updating",true);
            return cmp.getSuper().set("v.body",[]);
        }

        // No template, or no format index in value, use value as is
        var formatRegex=/\{(\d+)\}/gm;
        var template=cmp.get("v.template");
        if(!template||!template.length||!formatRegex.test(value)){
            // JBUCH: CHANGE TO $A.createComponent ONCE COMPONENTCONFIGS ARE AVAILABLE
            var valueText=$A.createComponentFromConfig({descriptor:"markup://aura:text",attributes:{value:value}});
            cmp.set("v.updating",true);
            return cmp.getSuper().set("v.body",[valueText]);
        }

        // Replace as many indexes as we can find, or as many body components, whichever comes first
        var body=[];
        var startIndex=0;
        value.replace(formatRegex, function(match, position, index) {
            var substitution = template[position];
            if (substitution !== undefined) {
                // JBUCH: CHANGE TO $A.createComponent ONCE COMPONENTCONFIGS ARE AVAILABLE
                body.push($A.createComponentFromConfig({descriptor:"markup://aura:text",attributes:{value:value.substring(startIndex,index)}}));
                body.push(substitution);
                startIndex=index+match.length;
            }
        });
        // Append the tail of the string
        if(startIndex<value.length){
            body.push($A.createComponentFromConfig({descriptor:"markup://aura:text",attributes:{value:value.substring(startIndex)}}));
        }
        cmp.set("v.updating",true);
        cmp.getSuper().set("v.body",body);
    }
})// eslint-disable-line semi
